h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamMojoModelBase.java

```diff
@@ -0,0 +1,175 @@
+  public double[] score0(double[] row, double[] preds) {
+    if (_meanImputation)
+      imputeMissingWithMeans(row);  // perform imputation for each row
+    
+    return gamScore0(row, preds);
+  }
+  
+  void init() {
+    _basisVals = new double[_gam_columns.length][];
+    _hj = new double[_gam_columns.length][];
+    for (int ind=0; ind < _num_gam_columns; ind++) {
+      _basisVals[ind] = new double[_num_knots[ind]];
+      _hj[ind] = ArrayUtils.eleDiff(_knots[ind]);
+    }
+    _lastClass = _nclasses - 1;
+  }
+  
+  abstract double[] gamScore0(double[] row, double[] preds);
+  
+  private void imputeMissingWithMeans(double[] data) {
+    for (int ind=0; ind < _cats; ind++)
+      if (Double.isNaN(data[ind])) data[ind] = _catNAFills[ind];
+
+    if (data.length == nfeatures()) { // using centered gam cols, nfeatures denotes centered gam columns
+      for (int ind = 0; ind < _numsCenter; ind++)
+        if (Double.isNaN(data[ind + _cats])) data[ind] = _numNAFillsCenter[ind];
+    } else {
+      for (int ind = 0; ind < _nums; ind++) {
+        int colInd = ind+_cats;
+        if (Double.isNaN(data[colInd])) 
+          data[colInd] = _numNAFills[ind];
+      }
+    }
+  }
+  
+  double evalLink(double val, LinkFunctionType linkType) {
+    switch (linkType) {
+      case identity: return GenModel.GLM_identityInv(val);
+      case logit: return GenModel.GLM_logitInv(val);
+      case log: return GenModel.GLM_logInv(val);
+      case inverse: return GenModel.GLM_inverseInv(val);
+      case tweedie: return GenModel.GLM_tweedieInv(val, _tweedieLinkPower);
+      default: throw new UnsupportedOperationException("Unexpected link function "+linkType);
+    }
+  }
+
+  // This method will read in categorical value and adjust for when useAllFactorLevels = true or false
+  int readCatVal(double data, int dataIndex) {
+    int ival = _useAllFactorLevels?((int) data):((int) data-1);
+    double targetVal = _useAllFactorLevels?(data):(data-1);
+    if (ival != targetVal) throw new IllegalArgumentException("categorical value out of range");
+    if (ival < 0)
+      return -1;
+    ival += _catOffsets[dataIndex];
+    return ival;
+  }
+
+  // this method will generate the beta*data+intercept
+  double generateEta(double[] beta, double[] data) {
+    double eta = 0.0;
+    for (int i = 0; i < _catOffsets.length - 1; ++i) {  // take care of contribution from categorical columns
+      int ival = readCatVal(data[i], i);
+      if ((ival < _catOffsets[i + 1]) && (ival >= 0))
+        eta += beta[ival];
+    }
+
+    int noff = _catOffsets[_cats] - _cats;
+    for (int i = _cats; i < beta.length - 1 - noff; ++i)
+      eta += beta[noff + i] * data[i];
+    eta += beta[beta.length - 1]; // add intercept
+    return eta;
+  }
+  
+  // this method will add to each data row the expanded gam columns
+  public void addExpandGamCols(double[] rawData, RowData rowData) { // add all expanded gam columns here
+    // add expanded gam columns to rowData
```

Comment:
mentions rowData but the method modifies rawData, I am confused

Reply:
@honzasterba for GAM, we need to take a predictor column, expanded it into basis functions and obtained the coefficients.  The rowData contains the value of the predictor column.  The newly expanded gam columns and the necessary predictor values are put into rawData which will later be used to do the scoring.  If you want to understand how the gam columns are constructed, I will slack you do doc I wrote.

Reply:
let me try to rephrase, the comment says "add expanded gam columns to rowData" yet I see no modifications done to rowData in the method, is the comment wrong or is the method wrong?
also the comment above the method mentions something about "each data row" yet this method operates strictly on one row of data
hence my confusion

Reply:
@honza:  I got what you mean now.

Okay, rowData contains say c0, c1, c2.  The user specify gam_columns = ["c0"].

So, the next step is Gam will take the value of c0 in rowData and generate a new set of gam_columns and put these new columns into rawData.  So, we need the information from rowData and modifies only rawData.

I can make RowData final though since it is not modified.  Thanks.